scipy/optimize/_minpack_py.py

```diff
@@ -800,8 +816,10 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
         raise ValueError("Method 'lm' only works for unconstrained problems. "
                          "Use 'trf' or 'dogbox' instead.")
 
-    # optimization may produce garbage for float32 inputs, cast them to float64
+    if check_finite is None:
+        check_finite = False if nan_policy is not None else True
```

Comment:
My fault; I think this would be easier to read, right?
```suggestion
        check_finite = True if nan_policy is None else False
```